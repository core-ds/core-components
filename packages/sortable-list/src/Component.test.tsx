import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import { SortableList, SortableListProps } from '.';

const items = [{ id: 1 }];
const testId = 'sortable';
const containerTestId = `${testId}-container`;
const itemTestId = `${testId}-item`;

const SortableListWrapper = (props?: Partial<SortableListProps>) => {
    return (
        <SortableList
            {...props}
            dataTestId={testId}
            items={items}
            renderItem={(item) => <div>{item.id}</div>}
        />
    );
};

describe('ScrollableList', () => {
    describe('Display tests', () => {
        it('should display correctly', () => {
            const { container } = render(<SortableListWrapper />);

            expect(container).toMatchSnapshot();
        });
    });

    describe('Props tests', () => {
        it('should set className on container', async () => {
            const containerClassName = 'container';
            const itemClassName = 'itemClassName';
            const { getByTestId } = render(
                <SortableListWrapper
                    className={containerClassName}
                    itemClassName={itemClassName}
                />,
            );

            const container = getByTestId(containerTestId);
            const item = getByTestId(itemTestId + '_1');

            expect(item).toHaveClass(itemClassName);
            expect(container).toHaveClass(containerClassName);
        });

        it('should set itemContentClassName on item content wrapper', async () => {
            const itemContentClassName = 'itemContentClassName';
            const { getByTestId } = render(
                <SortableListWrapper itemContentClassName={itemContentClassName} />,
            );

            const item = getByTestId(itemTestId + '_1');

            expect(item.firstElementChild).toHaveClass(itemContentClassName);
        });

        it('should set paddings as obj prop', async () => {
            const { getByTestId } = render(
                <SortableListWrapper
                    padding={{ top: '3xs', right: 'm', bottom: '2xs', left: 's' }}
                />,
            );

            const container = getByTestId(containerTestId);

            expect(container).toHaveClass(
                'padding-top-3xs',
                'padding-right-m',
                'padding-bottom-2xs',
                'padding-left-s',
            );
        });

        it('should set paddings as string prop', async () => {
            const { getByTestId } = render(<SortableListWrapper padding='m' />);

            const container = getByTestId(containerTestId);

            expect(container).toHaveClass(
                'padding-top-m',
                'padding-right-m',
                'padding-bottom-m',
                'padding-left-m',
            );
        });

        it('should set view', async () => {
            const { getByTestId } = render(<SortableListWrapper view='secondary' />);

            const container = getByTestId(containerTestId);
            const item = getByTestId(itemTestId + '_1');

            expect(container).toHaveClass('secondary');
            expect(item).toHaveClass('secondary');
        });

        it('should call onDragStart', async () => {
            const onDragStart = jest.fn();

            const { getByTestId } = render(<SortableListWrapper onDragStart={onDragStart} />);

            const item = getByTestId(itemTestId + '_1');

            await act(async () => {
                fireEvent.mouseDown(item);
            });

            expect(onDragStart).toBeCalledTimes(1);
        });

        it('should call onDragEnd', async () => {
            const onDragEnd = jest.fn();

            const { getByTestId } = render(<SortableListWrapper onDragEnd={onDragEnd} />);

            const item = getByTestId(itemTestId + '_1');

            await act(async () => {
                fireEvent.mouseDown(item);
            });

            await act(async () => {
                fireEvent.mouseUp(item);
            });

            expect(onDragEnd).toBeCalledTimes(1);
            expect(onDragEnd).toBeCalledWith(expect.anything(), items);
        });

        it('should set border-radius', async () => {
            const { getByTestId } = render(<SortableListWrapper borderRadius='m' />);

            const container = getByTestId(containerTestId);
            const item = getByTestId(itemTestId + '_1');

            expect(container).toHaveClass('border-radius-m');
            expect(item).toHaveClass('border-radius-m');
        });

        it('should set activator class', async () => {
            const { getByTestId } = render(<SortableListWrapper activatorNode='control' />);

            const item = getByTestId(itemTestId + '_1');

            expect(item).toHaveClass('activator-control');
        });

        it('should set control-padding-right class', () => {
            const { getByTestId } = render(<SortableListWrapper controlPadding='s' />);

            const item = getByTestId(itemTestId + '_1');
            const icon = item.childNodes[1];

            expect(icon).toHaveClass('control-right-padding-s');
        });

        it('should set control-padding-left class', () => {
            const { getByTestId } = render(
                <SortableListWrapper controlPadding='s' controlPosition='left' />,
            );

            const item = getByTestId(itemTestId + '_1');
            const icon = item.childNodes[0];

            expect(icon).toHaveClass('control-left-padding-s');
        });
    });

    describe('Render tests', () => {
        it('should unmount without errors', () => {
            const { unmount } = render(<SortableListWrapper />);

            expect(unmount).not.toThrowError();
        });
    });
});
