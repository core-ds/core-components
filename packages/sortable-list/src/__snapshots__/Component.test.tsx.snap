// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollableList Display tests should display correctly 1`] = `
<div>
  <div
    class="component primary"
    data-test-id="sortable-container"
  >
    <div
      aria-describedby="DndDescribedBy-0"
      aria-disabled="false"
      aria-roledescription="sortable"
      class="component activator-cell primary"
      data-test-id="sortable-item_1"
      role="button"
      tabindex="0"
    >
      <div
        class="content"
      >
        <div>
          1
        </div>
      </div>
      <div
        class="icon"
      >
        <svg
          aria-label="drag control"
          fill="currentColor"
          focusable="false"
          height="24"
          role="img"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            d="M8.5 7a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm0 6.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm1.5 5a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zM15.5 7a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm1.5 5a1.5 1.5 0 11-3 0 1.5 1.5 0 013 0zm-1.5 8a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    id="DndDescribedBy-0"
    style="display: none;"
  >
    Чтобы выбрать перетаскиваемый элемент, нажмите пробел.
При перетаскивании используйте клавиши со стрелками для перемещения элемента.
Снова нажмите пробел, чтобы переместить элемент в его новое положение или нажмите escape для отмены.
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  />
</div>
`;
